### PR TITLE
Patch

### LINKED ISSUE #7: PATCH Endpoint - Replace color within a palette
The main difference between the PUT and PATCH method is that the PUT method uses the request URI to supply a modified version of the requested resource which replaces the original version of the resource whereas the PATCH method supplies a set of **instructions to modify the resource**. If the PATCH document is larger than the size of the new version of the resource sent by the PUT method then the PUT method is preferred.[1]

Since we're editing from one hex color to just another hex color, both with the same length and format - let's use PATCH (modify resource rather than PUT - replace).

### PR DESCRIPTION
**Pull request by Dunn-Austen**
### What's this PR do?
- Integrates the edits from the last PR (Endpoints - Dustin)
- Adds the second PATCH endpoint (for editing a palette)
- Adds the happy path test for said PATCH
- Adds commented out code for the FAILING sad path test for said PATCH
- Revises the previous sad test for the first Patch, and also comments that out

### Any background information?
- The PATCH for altering a single color in a palette serves to alter from one of to all of the palette properties if need be
- Our combined endpoints in app.js and the corresponding tests in app.test.js could be polished by clustering all of the endpoint types together. I didn't group them because it makes the PR file changes a little overwhelming with so much relocating of code blocks. I'll open an issue and we can treat it a low-priority chore.
- I haven't able to solve the mystery of the failing sad path tests for the PATCH requests. Despite having nearly identical code to other sad tests, it's not working. The response.status does indeed return a 404 when the database is presented with a query for a non-existent ID value, but the response.body comes back as an empty object { } with no error property to display the error string from the implementation code.

```
    it('should return a sad 404 code if the targeted palette is not found', async () => {
      const invalidTargetId = -675324;
      const response = await request(app).get(`/api/v1/palettes/${invalidTargetId}`);
    
      expect(response.status).toBe(404);
      console.log(response.body);

      //The response body is returning an empty object (so response.body.error = undefined)

      expect(response.body.error).toEqual(`Could not find a palette with id: ${invalidTargetId}`);
    });
```

I've tried both
```
const response = await request(app).get(`/api/v1/palettes/${invalidTargetId}`);
```
And
```
const response = await request(app).patch(`/api/v1/palettes/${invalidTargetId}`);
```

**POSSIBLE SOLUTIONS:**
- Instead of testing **response.body.error**, we can maybe look at a separate property on the massive response object for testing.
- We could perhaps manipulate the the .json() clause in the implementation code to somehow populate the response object with a body that isn't an empty object
- We could expect only the failed 404 status, without the need for a clarifying error string

### Where should the reviewer start?
- Please try to troubleshoot the commented out tests

### Pertinent Issues
- Partially addresses #18 
- Resolves #7 

### Screenshots
-n/a

### CODE REVIEW

File: app.test.js [hunk 1]
```diff
@@ -101,14 +100,61 @@ describe('Server', () => {
       expect(revisedProject).toEqual(expectedResult);
     });
```
Comment (Dunn-Austen) [inline app.test.js]: Below is one of the sad path tests (PATCH) with which I was having trouble.
Would it be alright to just expect the response status to be 404?

File: app.js [hunk 2]
```diff
@@ -71,12 +71,52 @@ app.patch('/api/v1/projects/:id', async (request, response) => {
...
+    response.status(500).json({error: 'internal server error' })
+  }
+});
+
+app.patch('/api/v1/palettes/:id', async (request, response) => {
+  const { id } = request.params;
+  const newPalette = request.body;
+  try {
+    const patchedPalette = await database('palettes').where('id', id)
+    .update(
```
Comment (Dustin-Mikusko) [inline app.js]: I haven't worked with a PATCH yet, but it looks like knex has commands for everything and makes it easy. This is nice

File: app.test.js [hunk 3]
```diff
@@ -101,14 +101,62 @@ describe('Server', () => {
...
 
-    it('should return a sad 404 code if the targeted project is not found', async () => {
-      const invalidTargetId = -24;
+    // it('should return a sad 404 code if the targeted project is not found', async () => {
+    //   const invalidTargetId = -24;
+    //
+    //   const response = await request(app).get(`/api/v1/palettes/${invalidTargetId}`);
```
Comment (Dustin-Mikusko) [inline app.test.js]: I think the solution is adding `.send({invalidTargetId}`. to the end of the response line

File: app.test.js [hunk 4]
```diff
@@ -101,14 +101,62 @@ describe('Server', () => {
...
+        project_id: newPalette.project_id
+      };
+
+      expect(response.status).toBe(201);
+      expect(revisedPalette).toEqual(expectedResult);
     });
+
+    // it('should return a sad 404 code if the targeted palette is not found', async () => {
+    //   const invalidTargetId = -675324;
+    //   const response = await request(app).get(`/api/v1/palettes/${invalidTargetId}`);
```
Comment (Dustin-Mikusko) [inline app.test.js]: Same thing here, adding the `.send()` at the end

File: app.test.js (hunk 3 above)
Comment (Dunn-Austen) [reply app.test.js]: That worked!

Comment (Dunn-Austen) [reply app.test.js]: I've pushed up the tweaks. Way to go, man

Comment (Dunn-Austen) [reply app.test.js]: Wow

File: app.test.js [hunk 5]
```diff
@@ -104,13 +104,61 @@ describe('Server', () => {
     it('should return a sad 404 code if the targeted project is not found', async () => {
       const invalidTargetId = -24;
 
-      const response = await request(app).get(`/api/v1/projects/${invalidTargetId}`);
+      const response = await request(app).get(`/api/v1/projects/${invalidTargetId}`).send(`${invalidTargetId}`);
```
Comment (Dustin-Mikusko) [inline app.test.js]: BOOM

File: app.js (hunk 2 above)
Comment (Dunn-Austen) [reply app.js]: You're the man

File: app.test.js (hunk 5 above)
Comment (Dunn-Austen) [reply app.test.js]: Nailed it immediately. I'm impressed

**Pull request merged**